CCR unable to open any proxy connections

**Elasticsearch version** (`bin/elasticsearch --version`): 7.15

**Plugins installed**: default plugins in the docker image

**JVM version** (`java -version`): docker image

**OS version** (`uname -a` if on a Unix-like system): docker image

**Description of the problem including expected versus actual behavior**:

We have 2 separated  ELK clusters (3 nodes each) deployed in 2 different locations in Openshift. 
To deploy we are using a helm (no operator yet).
We enabled the CCR between them. The CCR connection is going through a F5 LoadBalancer.
The cluster is configured with security and TSL enabled.

The cross cluster replication is working fine but sometimes we see the following exception:

`[2021-11-19T14:16:58,628][WARN ][o.e.t.TcpTransport       ] [elasticsearch-master-0] exception caught on transport layer [Netty4TcpChannel{localAddress=/XXX.XXX.XXX.XXX:47822, remoteAddress=elk-master.local.net/YYY.YYY.YYY.YYY:443, profile=default}], closing connection

io.netty.handler.codec.DecoderException: javax.net.ssl.SSLHandshakeException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target

 

Caused by: javax.net.ssl.SSLHandshakeException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target

 

[2021-11-19T14:16:58,629][WARN ][o.e.x.c.a.ShardFollowNodeTask] [elasticsearch-master-0] shard follow task encounter non-retryable error

java.lang.IllegalStateException: Unable to open any proxy connections to remote cluster [elasticsearch-cluster-2]
`

Now we do not know exactly what is going on since the CCR before this exception was working fine. Maybe the F5 LB sometimes closes the connections or maybe sometime we have some network issue, we are still investigating.
However after we get that exception the CCR is broken.

To solve this issue we just have to restart the POD without ANY modification and the CCR starts to work again:

`[2021-11-22T10:24:31,168][INFO ][o.e.x.c.a.ShardFollowTasksExecutor] [elasticsearch-master-0] [myindex-2021.11.18-000001][0] Starting to track leader shard [myindex-2021.11.18-000001][0]

[2021-11-22T10:24:31,558][INFO ][o.e.x.c.a.ShardFollowNodeTask] [elasticsearch-master-0] [myindex-2021.11.18-000001][0] following leader shard [myindex-2021.11.18-000001][0], follower global checkpoint=[-1], mapping version=[2], settings version=[1], aliases version=[1]
`

Now I do not know if this is a bug or intended but we would like to find a solution in order to not have this issues again. From my point of view ELK should retry to open the proxy connections automatically.

Here the CCR settings:

`{
  "persistent" : {
    "cluster" : {
      "remote" : {
        "${REMOTE_CLUSTER_NAME}" : {
          "mode": "proxy",
          "proxy_address": "${remote_route}",
          "transport.ping_schedule": "5s"
        }
      }
    }
  }
}
`
**Steps to reproduce**:

Don't really know how to replicate this issue.

--- a/server/src/main/java/org/elasticsearch/transport/NoSeedNodeLeftException.java
+++ b/server/src/main/java/org/elasticsearch/transport/NoSeedNodeLeftException.java
@@ -14,12 +14,20 @@ import org.elasticsearch.common.io.stream.StreamInput;
 import java.io.IOException;
 
 /**
- * Thrown after failed to connect to all seed nodes of the remote cluster.
+ * Thrown after completely failing to connect to any node of the remote cluster.
  */
 public class NoSeedNodeLeftException extends ElasticsearchException {
 
-    public NoSeedNodeLeftException(String clusterName) {
-        super("no seed node left for cluster: [" + clusterName + "]");
+    public NoSeedNodeLeftException(String message) {
+        super(message);
+    }
+
+    NoSeedNodeLeftException(RemoteConnectionStrategy.ConnectionStrategy connectionStrategy, String clusterName) {
+        super(
+            connectionStrategy == RemoteConnectionStrategy.ConnectionStrategy.SNIFF
+                ? "no seed node left for cluster: [" + clusterName + "]"
+                : "Unable to open any proxy connections to cluster [" + clusterName + "]"
+        );
     }
 
     public NoSeedNodeLeftException(StreamInput in) throws IOException {

--- a/server/src/main/java/org/elasticsearch/transport/ProxyConnectionStrategy.java
+++ b/server/src/main/java/org/elasticsearch/transport/ProxyConnectionStrategy.java
@@ -287,9 +287,7 @@ public class ProxyConnectionStrategy extends RemoteConnectionStrategy {
         } else {
             int openConnections = connectionManager.size();
             if (openConnections == 0) {
-                finished.onFailure(
-                    new IllegalStateException("Unable to open any proxy connections to remote cluster [" + clusterAlias + "]")
-                );
+                finished.onFailure(new NoSeedNodeLeftException(strategyType(), clusterAlias));
             } else {
                 logger.debug(
                     "unable to open maximum number of connections [remote cluster: {}, opened: {}, maximum: {}]",

--- a/server/src/main/java/org/elasticsearch/transport/SniffConnectionStrategy.java
+++ b/server/src/main/java/org/elasticsearch/transport/SniffConnectionStrategy.java
@@ -339,7 +339,7 @@ public class SniffConnectionStrategy extends RemoteConnectionStrategy {
                 onFailure.accept(e);
             });
         } else {
-            listener.onFailure(new NoSeedNodeLeftException(clusterAlias));
+            listener.onFailure(new NoSeedNodeLeftException(strategyType(), clusterAlias));
         }
     }
 

--- a/server/src/test/java/org/elasticsearch/transport/ProxyConnectionStrategyTests.java
+++ b/server/src/test/java/org/elasticsearch/transport/ProxyConnectionStrategyTests.java
@@ -34,6 +34,9 @@ import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.function.Supplier;
 import java.util.stream.Collectors;
 
+import static org.hamcrest.Matchers.allOf;
+import static org.hamcrest.Matchers.containsString;
+
 public class ProxyConnectionStrategyTests extends ESTestCase {
 
     private final String clusterAlias = "cluster-alias";
@@ -202,7 +205,10 @@ public class ProxyConnectionStrategyTests extends ESTestCase {
 
                     PlainActionFuture<Void> connectFuture = PlainActionFuture.newFuture();
                     strategy.connect(connectFuture);
-                    expectThrows(Exception.class, connectFuture::actionGet);
+                    assertThat(
+                        expectThrows(NoSeedNodeLeftException.class, connectFuture::actionGet).getMessage(),
+                        allOf(containsString("Unable to open any proxy connections"), containsString('[' + clusterAlias + ']'))
+                    );
 
                     assertFalse(connectionManager.getAllConnectedNodes().stream().anyMatch(n -> n.getAddress().equals(address1)));
                     assertEquals(0, connectionManager.size());

--- a/x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java
+++ b/x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java
@@ -632,13 +632,10 @@ public abstract class ShardFollowNodeTask extends AllocatedPersistentTask {
             || actual instanceof NoShardAvailableActionException
             || actual instanceof UnavailableShardsException
             || actual instanceof AlreadyClosedException
-            || actual instanceof ElasticsearchSecurityException
-            || // If user does not have sufficient privileges
-            actual instanceof ClusterBlockException
-            || // If leader index is closed or no elected master
-            actual instanceof IndexClosedException
-            || // If follow index is closed
-            actual instanceof ConnectTransportException
+            || actual instanceof ElasticsearchSecurityException // If user does not have sufficient privileges
+            || actual instanceof ClusterBlockException // If leader index is closed or no elected master
+            || actual instanceof IndexClosedException // If follow index is closed
+            || actual instanceof ConnectTransportException
             || actual instanceof NodeClosedException
             || actual instanceof NoSuchRemoteClusterException
             || actual instanceof NoSeedNodeLeftException

